pkg/provider/cloud/gcp/firewall_rules.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcp
+
+import (
+	"fmt"
+	"net/http"
+
+	"go.uber.org/zap"
+	"google.golang.org/api/compute/v1"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/provider"
+	"k8c.io/kubermatic/v2/pkg/resources"
+)
+
+func reconcileFirewallRules(cluster *kubermaticv1.Cluster, update provider.ClusterUpdater, svc *compute.Service, projectID string) error {
+	// Retrieve nodePort range from cluster
+	nodePortRangeLow, nodePortRangeHigh := resources.NewTemplateDataBuilder().
+		WithNodePortRange(cluster.Spec.ComponentsOverride.Apiserver.NodePortRange).
+		WithCluster(cluster).
+		Build().
+		NodePorts()
+
+	nodePortsAllowedIPRange := cluster.Spec.Cloud.GCP.NodePortsAllowedIPRange
+	if nodePortsAllowedIPRange == "" {
+		nodePortsAllowedIPRange = "0.0.0.0/0"
+	}
+
+	firewallService := compute.NewFirewallsService(svc)
+	tag := fmt.Sprintf("kubernetes-cluster-%s", cluster.Name)
+	selfRuleName := fmt.Sprintf("firewall-%s-self", cluster.Name)
+	icmpRuleName := fmt.Sprintf("firewall-%s-icmp", cluster.Name)
+	nodePortRuleName := fmt.Sprintf("firewall-%s-nodeport", cluster.Name)
+
+	// allow traffic within the same cluster
+	if !kuberneteshelper.HasFinalizer(cluster, firewallSelfCleanupFinalizer) {
```

Comment:
This condition means that once the firewall rules were created, the provider will not attempt to reconcile them on further runs. The new reconcile functionality is supposed to find resources that went missing from the cloud provider and re-create them if needed. I think this needs to be replaced with a query to GCP to check if the firewall service is there.